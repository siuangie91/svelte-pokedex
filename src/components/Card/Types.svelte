<script lang="ts">
  import type { Type } from 'src/types';
  import { capitalizeFirstLetter, TYPE_COLOR_LOOKUP } from 'utils';

  export let types: Type[] = [];
</script>

{#if types.length}
  <h2>Types</h2>
  <ul style="--num-types: {types.length}">
    {#each types as { type }}
      <li
        style="--type-bg-color: {TYPE_COLOR_LOOKUP[type.name].color}; --type-text-color: {TYPE_COLOR_LOOKUP[type.name]
          .dark
          ? '#fff'
          : '#000'}"
      >
        {capitalizeFirstLetter(type.name)}
      </li>
    {/each}
  </ul>
{/if}

<style lang="postcss">
  h2 {
    @apply text-lg;
  }

  ul {
    @apply grid;
    @apply gap-4;
    grid-template-columns: repeat(var(--num-types), minmax(0, 1fr));
  }

  li {
    @apply p-1;
    @apply rounded-sm;
    @apply text-sm;
    @apply text-center;
    background-color: var(--type-bg-color);
    color: var(--type-text-color);
  }
</style>
