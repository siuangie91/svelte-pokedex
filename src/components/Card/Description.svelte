<script lang="ts">
  export let description: string = '';
</script>

{#if description}
  <h2 class="card-h2">Description</h2>
  <p class="description">{description}</p>
{/if}

<style lang="postcss">
  .description {
    @apply text-base;
  }
</style>
