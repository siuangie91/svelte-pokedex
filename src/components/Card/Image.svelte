<script lang="ts">
  export let image: string | null = null;
  export let name: string = '';
</script>

{#if image}
  <img src={image} alt={name} />
{/if}

<style lang="postcss">
  img {
    @apply block;
    @apply w-3/4;
    @apply mx-auto;
  }
</style>
