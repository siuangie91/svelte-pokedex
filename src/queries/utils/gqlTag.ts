/**
 * Workaround for error when running `vite build`
 * `TypeError: gql is not a function`
 * https://github.com/vitejs/vite/issues/5694#issuecomment-976098129
 */
import GqlTag from 'gql-tag';
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const gql = (GqlTag as any).default || GqlTag;

export default gql;
