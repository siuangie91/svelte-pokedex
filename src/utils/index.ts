// TODO test
import axios from 'axios';
import type { Pokemon } from 'src/types';

const POKEAPI_GEN_1_URL = 'https://pokeapi.co/api/v2/pokedex/1';

/**
 * Fetches the first 20 entries for Gen 1 pokemon from PokeAPI
 * @returns
 */
export const getFirst20PokemonEntries = async () => {
  try {
    const { data } = await axios.get<PokemonAPI.ApiResponse>(POKEAPI_GEN_1_URL);

    const { pokemon_entries } = data;

    return pokemon_entries.length ? pokemon_entries.slice(0, 20) : [];
  } catch (error) {
    console.error(error);
  }
};

/**
 * Takes the values `pokemon_entries` and returns
 * an array of flattened `Pokemon` objects
 * @param entries
 * @returns
 */
export const flattenEntries = (entries: PokemonAPI.Entry[]): Pokemon[] => {
  return entries.map(({ entry_number, pokemon_species }) => {
    return {
      id: entry_number,
      name: pokemon_species.name,
      url: pokemon_species.url,
    };
  });
};

/**
 * Creates a lookup table of Pokemon by name
 * @param entries Pokemon entrees
 * @returns
 */
export const createLookupByName = (entries: Pokemon[]) => {
  return entries.reduce<{
    [key: string]: Pokemon;
  }>((acc, { id, name, url }) => {
    acc[name] = {
      id,
      name,
      url,
    };

    return acc;
  }, {});
};

/**
 * Capitalizes the first letter of a string
 * @param str
 * @returns
 */
export const capitalizeFirstLetter = (str: string) => {
  const firstLetter = str[0];
  const rest = str.slice(1);
  return `${firstLetter.toUpperCase()}${rest}`;
};
