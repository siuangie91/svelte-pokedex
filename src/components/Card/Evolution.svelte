<script lang="ts">
  import { capitalizeFirstLetter } from 'utils';

  export let name: string;
  export let evolutions: string[] = [];

  let numColumns = 0;
  if (evolutions.length <= 3) {
    numColumns = evolutions.length;
  } else {
    numColumns = evolutions.length % 2 === 0 ? 4 : 3;
  }
</script>

{#if evolutions.length > 1}
  <h2>Evolution</h2>
  <ol
    style="
      --num-columns: {numColumns}
    "
  >
    {#each evolutions as evolution}
      <li style="--evolution-weight: {evolution.toLowerCase() === name.toLowerCase() ? 'bold' : 'normal'}">
        {#if evolution === name}
          <span>{capitalizeFirstLetter(evolution)}</span>
        {:else}
          <a sveltekit:reload href={`/pokemon/${evolution}`}>{capitalizeFirstLetter(evolution)}</a>
        {/if}
      </li>
    {/each}
  </ol>
{/if}

<style lang="postcss">
  h2 {
    @apply text-lg;
  }

  ol {
    @apply grid;
    @apply gap-4;
    @apply gap-y-1;
    grid-template-columns: repeat(var(--num-columns), minmax(0, 1fr));
  }

  li {
    @apply relative;
    @apply text-center;
    @apply text-sm;

    /* right arrow */
    @apply after:content-['\0279E'];
    @apply after:absolute;
    @apply after:-right-3.5;
    @apply after:font-normal;

    @apply last-of-type:after:content-none;

    font-weight: var(--evolution-weight);
  }
</style>
