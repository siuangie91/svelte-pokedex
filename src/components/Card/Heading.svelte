<script lang="ts">
  import { capitalizeFirstLetter } from 'utils';

  export let id: number;
  export let name: string;
</script>

{#if id && name}
  <h1>
    <span>{capitalizeFirstLetter(name)}</span>
    <span>#{id}</span>
  </h1>
{/if}

<style lang="postcss">
  h1 {
    @apply m-0;
    @apply text-xl;

    @apply flex;
    @apply justify-between;
  }
</style>
