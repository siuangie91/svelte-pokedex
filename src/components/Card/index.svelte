<script lang="ts">
  import type { Type } from 'src/types';
  import { capitalizeFirstLetter } from 'utils';

  export let id: number;
  export let name: string;
  export let image: string | null = null;
  export let types: Type[] = [];
  export let evolution: string[] = [];
  export let description: string = '';

  // const gridTemplateColumns = `repeat(${types.length}, minmax(0, 1fr))`;
</script>

<section>
  <div>
    <h1>{id}. {name}</h1>

    {#if image}
      <img src={image} alt={`${name}`} />
    {/if}

    {#if types.length}
      <h2 class="card-h2">Types</h2>
      <ul class="list-types" style="--num-types: {types.length}">
        {#each types as { type }}
          <li>{capitalizeFirstLetter(type.name)}</li>
        {/each}
      </ul>
    {/if}

    {#if evolution.length}
      <h2 class="card-h2">Evolution</h2>
      <ol class="list-evolution" style="--num-evolution: {evolution.length}">
        {#each evolution as name}
          <li>{capitalizeFirstLetter(name)}</li>
        {/each}
      </ol>
    {/if}

    {#if description}
      <h2 class="card-h2">Description</h2>
      <p class="description">{description}</p>
    {/if}
  </div>
</section>

<style lang="postcss">
  section {
    max-width: 350px;
    @apply mx-auto;
    @apply p-4;
    @apply bg-amber-300;
    @apply rounded-lg;
  }

  div {
    @apply p-4;
    @apply bg-white;
  }

  h1 {
    @apply m-0;
    @apply text-lg;
  }

  .card-h2 {
    @apply text-sm;
  }

  img {
    @apply block;
    @apply w-3/4;
    @apply mx-auto;
  }

  .list-types,
  .list-evolution {
    @apply grid;
    @apply gap-4;
  }

  .list-types {
    grid-template-columns: repeat(var(--num-types), minmax(0, 1fr));
  }
  .list-types li {
    background-color: gray;
    @apply rounded-sm;
    @apply text-center;
  }

  .list-evolution {
    grid-template-columns: repeat(var(--num-evolution), minmax(0, 1fr));
  }
  .list-evolution li {
    background-color: gray;
    @apply rounded-sm;
    @apply text-center;
    @apply relative;

    @apply after:content-['â†’'];
    @apply after:absolute;
    @apply after:-right-4;

    @apply last-of-type:after:content-none;
  }

  .description {
    @apply text-sm;
  }
</style>
