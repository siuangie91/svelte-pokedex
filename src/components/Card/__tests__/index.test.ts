import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/svelte';
import { capitalizeFirstLetter } from 'utils';
import type { Type } from 'src/types';
import Card from '../index.svelte';

describe('Card', () => {
  it('renders all the pokemon information like pokemon card', () => {
    const id = 1;
    const name = "bulbasaur";
    const image = "path/to/image";
    const types = [
      { type: { name: 'grass' } }, 
      { type: { name: 'grass' } },
    ];
    const evolutions = [name, 'ivysaur', 'venusaur'];
    const description = 'mock description';

    const { container } = render(Card, {
      id,
      name,
      image,
      types,
      evolutions,
      description
    });

    expect(container).toMatchSnapshot();
  });
});
