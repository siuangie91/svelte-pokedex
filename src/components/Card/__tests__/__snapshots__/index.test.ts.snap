// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders all the pokemon information like pokemon card 1`] = `
<body>
  <div>
    <section
      class="svelte-1t5s68u"
      style="--card-bg-color: #03ab78;"
    >
      <div
        class="svelte-1t5s68u"
      >
        <h1
          class="svelte-1ma9vo8"
        >
          <span>
            Bulbasaur
          </span>
           
          <span>
            #
            1
          </span>
        </h1>
        
         
        <img
          alt="bulbasaur"
          class="svelte-wr7qya"
          src="path/to/image"
        />
        
         
        <h2
          class="svelte-8hcvkc"
        >
          Types
        </h2>
         
        <ul
          class="svelte-8hcvkc"
          style="--num-types: 2;"
        >
          <li
            class="svelte-8hcvkc"
            style="--type-bg-color: #03ab78; --type-text-color: #000;"
          >
            Grass
             
          </li>
          <li
            class="svelte-8hcvkc"
            style="--type-bg-color: #03ab78; --type-text-color: #000;"
          >
            Grass
             
          </li>
        </ul>
        
         
        <h2
          class="svelte-13bn853"
        >
          Evolution
        </h2>
         
        <ol
          class="svelte-13bn853"
          style="--num-columns: 3;"
        >
          <li
            class="svelte-13bn853"
            style="--evolution-weight: bold;"
          >
            <span>
              Bulbasaur
            </span>
             
          </li>
          <li
            class="svelte-13bn853"
            style="--evolution-weight: normal;"
          >
            <a
              href="/pokemon/ivysaur"
              sveltekit:reload=""
            >
              Ivysaur
            </a>
             
          </li>
          <li
            class="svelte-13bn853"
            style="--evolution-weight: normal;"
          >
            <a
              href="/pokemon/venusaur"
              sveltekit:reload=""
            >
              Venusaur
            </a>
             
          </li>
        </ol>
        
         
        <h2
          class="card-h2"
        >
          Description
        </h2>
         
        <p
          class="description svelte-185rl9w"
        >
          mock description
        </p>
        
      </div>
    </section>
  </div>
</body>
`;
